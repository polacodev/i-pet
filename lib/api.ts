import { decode } from 'base64-arraybuffer';

import { supabase } from '@/lib/supabase';
import { PetProps } from '@/types/pet';
import { Session } from '@/types/session';
import { UserProfileProps } from '@/types/user';

export const getPetList = async (userId: string | null) => {
  let petList: PetProps[] = [];
  const { data, error } = await supabase
    .from('pets')
    .select(
      `
      id,
      inserted_at,
      pet_image,
      pet_age,
      pet_breed,
      pet_gender,
      pet_medical_condition,
      pet_name,
      pet_type,
      user_id,
      profiles (
        id,
        email,
        country_code,
        phone
      )
      `,
    )
    .eq('user_id', userId);

  // if (error) return { data, error };
  // if (data) {
  petList = data;
  return { data: petList, error };
  // }
  // return { data, error };
};

export const getPetById = async (petId: string) => {
  const { data, error } = await supabase
    .from<any, 'public'>('pets')
    .select(
      `
      id,
      inserted_at,
      pet_image,
      pet_age,
      pet_breed,
      pet_gender,
      pet_medical_condition,
      pet_name,
      pet_type,
      user_id,
      profiles (
        id,
        email,
        country_code,
        phone
      )
    `,
    )
    .eq('id', petId)
    .single();
  return { data, error };
};

export const deletePetById = async (petId: string) => {
  const { error } = await supabase.from('pets').delete().eq('id', petId);
  return { error };
};

export const insertPet = async (newPet: any) => {
  const { error } = await supabase.from('pets').upsert(newPet);
  return { error };
};

export const storePetImage = async (filePath: string, code: any, contentType: string) => {
  const { data, error } = await supabase.storage
    .from('pets')
    .upload(filePath, decode(code), { contentType });
  return { data, error };
};

export const deletePetImage = async (petImage: string) => {
  const { data, error } = await supabase.storage.from('pets').remove([petImage]);
  return { data, error };
};

export const signInWithPassword = async (email: string, password: string) => {
  const { data, error } = await supabase.auth.signInWithPassword({
    email,
    password,
  });
  return { data, error };
};

export const signUpWithEmailPassword = async (email: string, password: string) => {
  const { data, error } = await supabase.auth.signUp({ email, password });
  return { data, error };
};

export const saveUserProfile = async (user: UserProfileProps) => {
  const { data, error } = await supabase.from('profiles').upsert(user);
  return { data, error };
};

export const signOut = async () => {
  await supabase.auth.signOut();
};

export const getSession = async () => {
  return await supabase.auth.getSession();
};

export const onAuthStateChange = (setSession: (session: Session | null) => void) => {
  return supabase.auth.onAuthStateChange((_event, session) => {
    setSession(session);
  });
};
