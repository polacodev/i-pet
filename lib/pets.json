[
  {
    "id": "550e8400-e29b-41d4-a716-446655440000",
    "petName": "Bella",
    "petType": "Dog",
    "petAge": 3,
    "petBreed": "Golden Retriever",
    "petGender": "Female",
    "petMedicalCondition": "Hip Dysplasia",
    "ownerName": "John Smith",
    "ownerPhone": "60717546"
  },
  {
    "id": "550e8400-e29b-41d4-a716-446655440001",
    "petName": "Max",
    "petType": "Dog",
    "petAge": 2,
    "petBreed": "Bulldog",
    "petGender": "Male",
    "petMedicalCondition": "Allergies",
    "ownerName": "Emily Johnson",
    "ownerPhone": "60717546"
  },
  {
    "id": "550e8400-e29b-41d4-a716-446655440002",
    "petName": "Luna",
    "petType": "Cat",
    "petAge": 1,
    "petBreed": "Siamese",
    "petGender": "Female",
    "petMedicalCondition": "Asthma",
    "ownerName": "Michael Brown",
    "ownerPhone": "60717546"
  },
  {
    "id": "550e8400-e29b-41d4-a716-446655440003",
    "petName": "Charlie",
    "petType": "Dog",
    "petAge": 5,
    "petBreed": "Labrador",
    "petGender": "Male",
    "petMedicalCondition": "Arthritis",
    "ownerName": "Jessica Davis",
    "ownerPhone": "60717546"
  },
  {
    "id": "550e8400-e29b-41d4-a716-446655440004",
    "petName": "Daisy",
    "petType": "Dog",
    "petAge": 4,
    "petBreed": "Beagle",
    "petGender": "Female",
    "petMedicalCondition": "Diabetes",
    "ownerName": "David Wilson",
    "ownerPhone": "60717546"
  }
]