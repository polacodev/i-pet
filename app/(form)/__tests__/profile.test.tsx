import { fireEvent, render, waitFor } from '@testing-library/react-native';
import '@testing-library/react-native/extend-expect';
import { router } from 'expo-router';
import React from 'react';

import Profile from '@/app/(form)/profile';

jest.mock('expo-router', () => ({
  router: {
    navigate: jest.fn(),
  },
}));

jest.mock('@/localizations/localization', () => ({
  localization: {
    t: jest.fn().mockImplementation((key) => key),
  },
}));

jest.mock('@/lib/api', () => ({
  signOut: jest.fn(),
  getSession: jest.fn(),
  onAuthStateChange: jest.fn(),
}));

describe('<Profile />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    (router.navigate as jest.Mock).mockClear();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test('Profile renders correctly and Snapshot', async () => {
    require('@/lib/api').getSession.mockReturnValue({
      data: {
        session: {
          user: {
            email: 'test@example.com',
          },
        },
      },
    });

    const setSessionStateMock = jest.fn();
    const useStateMock: any = (useState: any) => [useState, setSessionStateMock];

    const useEffectSpy = jest.spyOn(React, 'useEffect');
    const useStateSpy = jest.spyOn(React, 'useState').mockImplementation(useStateMock);

    const { toJSON } = render(<Profile />);

    await waitFor(() => {
      expect(useEffectSpy).toHaveBeenCalled();
      expect(useStateSpy).toHaveBeenCalled();
    });

    expect(toJSON()).toMatchSnapshot();
  });

  test('onSignOut event is called when log-out icon is pressed', async () => {
    require('@/lib/api').getSession.mockReturnValue({
      data: {
        session: {
          user: {
            email: 'test@example.com',
          },
        },
      },
    });

    const { getByTestId } = render(<Profile />);

    fireEvent.press(getByTestId('pet-button'), 'onClick');

    await waitFor(() => {
      expect(router.navigate).toHaveBeenCalledWith('/');
    });
  });
});
