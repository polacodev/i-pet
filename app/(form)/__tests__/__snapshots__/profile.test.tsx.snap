// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile /> Profile renders correctly and Snapshot 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#f6f6f6",
      },
      {
        "width": "100%",
      },
    ]
  }
  testID="pet-view"
>
  <View
    style={
      [
        {
          "backgroundColor": "#f6f6f6",
        },
        {
          "alignItems": "center",
          "gap": 5,
          "justifyContent": "center",
        },
      ]
    }
    testID="pet-view"
  >
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          {
            "paddingHorizontal": 50,
            "paddingTop": 20,
          },
        ]
      }
      testID="pet-text"
    >
      header_log_out_message
    </Text>
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 14,
            "lineHeight": 20,
          },
          {
            "color": "#6b7280",
          },
        ]
      }
      testID="pet-text"
    >
      test@example.com
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#0e7490",
          "borderRadius": 10,
          "flexDirection": "row",
          "gap": 8,
          "justifyContent": "center",
          "margin": 6,
          "opacity": 1,
          "padding": 8,
          "width": "85%",
        }
      }
      testID="pet-button"
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#ffffff",
              "fontSize": 30,
            },
            [
              undefined,
            ],
            {
              "fontFamily": "ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
        testID="pet-icon"
      >
        î³‘
      </Text>
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            {
              "color": "#ffffff",
            },
          ]
        }
        testID="pet-text"
      >
        header_log_out
      </Text>
    </View>
  </View>
</View>
`;
